File: python/tests/test_inexact_predictor.py
"""
Validation tests for inexact predictor feedback.
Fang & Zhang (2024), Section 6.
"""

import sys
import os
import numpy as np
import pytest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from src.inexact_predictor import run_inexact_predictor


class TestConvergence:
    """test_convergence -- State converges to zero under inexact predictor."""

    def test_state_converges_to_zero(self):
        """State norm at t_end should be below 0.1."""
        res = run_inexact_predictor()
        final_norm = np.linalg.norm(res['x_hist'][-1])
        print(f'    |x(t_end)| = {final_norm:.4e}')
        assert final_norm < 0.1, \
            f'State did not converge: |x(t_end)| = {final_norm:.4e}'

    def test_x1_converges(self):
        """x1 component should converge to zero."""
        res = run_inexact_predictor()
        assert abs(res['x_hist'][-1, 0]) < 0.1, \
            f'x1 did not converge: x1(t_end) = {res["x_hist"][-1, 0]:.4e}'

    def test_x2_converges(self):
        """x2 component should converge to zero."""
        res = run_inexact_predictor()
        assert abs(res['x_hist'][-1, 1]) < 0.1, \
            f'x2 did not converge: x2(t_end) = {res["x_hist"][-1, 1]:.4e}'


class TestPredictorAtT0:
    """test_predictor_at_t0 -- Verify predictor ODE at t=0 vs independent integration."""

    def test_predictor_matches_independent(self):
        """Predictor at t=0 should match independent forward integration."""
        D0  = 0.45
        x0  = np.array([1.0, 1.0])
        dt  = 0.001

        N_pred = int(round(D0 / dt))
        d_theta = D0 / N_pred

        # At t=0, the predictor integrates from theta = -D0 to 0.
        # All past controls are 0 (pre-history).
        P1, P2 = x0[0], x0[1]
        for j in range(N_pred):
            u1_theta = 0.0  # pre-history
            u2_theta = 0.0  # pre-history
            dP1 = P2
            dP2 = -P1 + np.sin(P1) + P2 + u1_theta + np.tanh(u1_theta) + np.sin(u2_theta)
            P1 = P1 + d_theta * dP1
            P2 = P2 + d_theta * dP2

        P_manual = np.array([P1, P2])

        # Run full simulation and extract P(0)
        res = run_inexact_predictor({'x0': x0.tolist(), 'dt': dt, 'D0': D0})
        P_sim = res['P_hist'][0, :]

        err = np.linalg.norm(P_manual - P_sim)
        print(f'    P_manual = [{P_manual[0]:.6f}, {P_manual[1]:.6f}]')
        print(f'    P_sim    = [{P_sim[0]:.6f}, {P_sim[1]:.6f}]')
        print(f'    |error|  = {err:.4e}')

        assert err < 1e-10, \
            f'Predictor at t=0 does not match independent integration: err = {err:.4e}'


class TestControlInitial:
    """test_control_initial -- First control values are correct."""

    def test_first_u1_correct(self):
        """u1(0) should match -(kp1*P1(0) + kd1*P2(0))."""
        res = run_inexact_predictor()
        kp1 = 3.0
        kd1 = 5.0
        P1_0 = res['P_hist'][0, 0]
        P2_0 = res['P_hist'][0, 1]
        u1_expected = -(kp1 * P1_0 + kd1 * P2_0)
        u1_actual = res['u1_hist'][0]
        err = abs(u1_expected - u1_actual)
        print(f'    u1(0) expected = {u1_expected:.6f}')
        print(f'    u1(0) actual   = {u1_actual:.6f}')
        print(f'    |error|        = {err:.4e}')
        assert err < 1e-10, f'u1(0) mismatch: err = {err:.4e}'

    def test_first_u2_correct(self):
        """u2(0) should match -(kp2*P1(0) + kd2*P2(0))."""
        res = run_inexact_predictor()
        kp2 = 10.0
        kd2 = 10.0
        P1_0 = res['P_hist'][0, 0]
        P2_0 = res['P_hist'][0, 1]
        u2_expected = -(kp2 * P1_0 + kd2 * P2_0)
        u2_actual = res['u2_hist'][0]
        err = abs(u2_expected - u2_actual)
        print(f'    u2(0) expected = {u2_expected:.6f}')
        print(f'    u2(0) actual   = {u2_actual:.6f}')
        print(f'    |error|        = {err:.4e}')
        assert err < 1e-10, f'u2(0) mismatch: err = {err:.4e}'


class TestDtRefinement:
    """test_dt_refinement -- O(dt) convergence order estimate."""

    def test_convergence_order(self):
        """Average convergence order should be at least 0.8."""
        params_base = {'x0': [1.0, 1.0], 't_end': 5.0}

        dt_vals = [0.004, 0.002, 0.001, 0.0005]

        # Reference: dt = 0.00025
        params_ref = {**params_base, 'dt': 0.00025}
        res_ref = run_inexact_predictor(params_ref)

        errors = np.zeros(len(dt_vals))
        for i, dt_i in enumerate(dt_vals):
            params_i = {**params_base, 'dt': dt_i}
            res_i = run_inexact_predictor(params_i)

            # Interpolate reference onto this grid
            x_ref_interp = np.zeros_like(res_i['x_hist'])
            for col in range(2):
                x_ref_interp[:, col] = np.interp(
                    res_i['t'], res_ref['t'], res_ref['x_hist'][:, col])
            errors[i] = np.max(np.abs(res_i['x_hist'] - x_ref_interp))
            print(f'    dt = {dt_i:.5f}: max error = {errors[i]:.4e}')

        # Compute convergence orders between successive refinements
        print('    Convergence orders:')
        orders = []
        for i in range(len(dt_vals) - 1):
            if errors[i + 1] > 1e-14:
                ratio = dt_vals[i] / dt_vals[i + 1]
                order = np.log(errors[i] / errors[i + 1]) / np.log(ratio)
                orders.append(order)
                print(f'      dt {dt_vals[i]:.5f} -> {dt_vals[i+1]:.5f}: '
                      f'order = {order:.2f}')

        if orders:
            avg_order = np.mean(orders)
            print(f'    Average order = {avg_order:.2f} (expected ~1.0)')
            assert avg_order > 0.8, \
                f'Average convergence order too low: {avg_order:.2f}'


class TestExponentialDecay:
    """test_exponential_decay -- Negative slope of log(norm(x)) after transient."""

    def test_negative_slope(self):
        """log(||x||) should have a negative slope after initial transient."""
        res = run_inexact_predictor()
        t = res['t']
        x_norm = np.linalg.norm(res['x_hist'], axis=1)

        # Use data after initial transient (t > 2.0)
        mask = t > 2.0
        t_tail = t[mask]
        x_norm_tail = x_norm[mask]

        # Remove any near-zero values to avoid log issues
        valid = x_norm_tail > 1e-15
        t_valid = t_tail[valid]
        log_norm = np.log(x_norm_tail[valid])

        # Linear fit: log(||x||) ~ a*t + b
        if len(t_valid) > 10:
            coeffs = np.polyfit(t_valid, log_norm, 1)
            slope = coeffs[0]
            print(f'    Slope of log(||x||) for t > 2.0: {slope:.4f}')
            assert slope < -0.1, \
                f'Exponential decay not observed: slope = {slope:.4f}'
        else:
            pytest.skip('Not enough valid data points for exponential decay test.')

File: matlab/tests/test_inexact_predictor.m
function tests = test_inexact_predictor
%TEST_INEXACT_PREDICTOR  Unit tests for the inexact predictor controller.
%
%  Tests convergence, predictor initialisation, control computation,
%  time-step refinement order, and exponential decay of the state norm.
%
%  Run with:  results = runtests('test_inexact_predictor');

    tests = functiontests(localfunctions);
end

%% -----------------------------------------------------------------------
%  test_convergence
%  -----------------------------------------------------------------------
%  After t_end = 10 the state should be near the origin.
function test_convergence(testCase)
    fprintf('\n--- test_convergence ---\n');

    p.dt    = 0.001;
    p.t_end = 10.0;
    r = run_inexact_predictor(p);

    final_norm = norm(r.x_hist(end,:));
    fprintf('  Final state norm: %.6e\n', final_norm);

    verifyLessThan(testCase, final_norm, 0.1, ...
        'State should converge close to the origin by t=10.');
end

%% -----------------------------------------------------------------------
%  test_predictor_at_t0
%  -----------------------------------------------------------------------
%  At t = 0 the predictor integrates from theta = -D0 to 0.
%  Because all control pre-history is zero, we can reproduce this
%  integration independently.  P(−D0) = x0 = [1;1], u1=u2=0,
%  dynamics: dP1 = P2, dP2 = −P1 + sin(P1) + P2.
function test_predictor_at_t0(testCase)
    fprintf('\n--- test_predictor_at_t0 ---\n');

    D0  = 0.45;
    dt  = 0.001;
    x0  = [1; 1];

    % Run the full simulation (only need step k=1 result)
    p.dt    = dt;
    p.t_end = 0.1;       % short run, we only need P_hist(1,:)
    p.D0    = D0;
    r = run_inexact_predictor(p);

    P_sim = r.P_hist(1,:);   % predictor output at t=0
    fprintf('  P_hist(1,:) from simulation: [%.10f, %.10f]\n', P_sim(1), P_sim(2));

    % Manual integration: Euler, same dt, u1=u2=0 everywhere
    N_pred = round(D0 / dt);
    P1 = x0(1);
    P2 = x0(2);
    for j = 0:N_pred-1
        dP1 = P2;
        dP2 = -P1 + sin(P1) + P2;   % u1=u2=0 -> tanh(0)=0, sin(0)=0
        P1 = P1 + dt * dP1;
        P2 = P2 + dt * dP2;
    end
    P_manual = [P1, P2];
    fprintf('  Manual integration result:   [%.10f, %.10f]\n', P_manual(1), P_manual(2));

    err = norm(P_sim - P_manual);
    fprintf('  Error: %.4e\n', err);

    verifyLessThan(testCase, err, 1e-10, ...
        'Predictor at t=0 should match manual Euler integration exactly.');
end

%% -----------------------------------------------------------------------
%  test_control_initial
%  -----------------------------------------------------------------------
%  At k=1 the control is computed from the predictor output P:
%    u1(0) = -(kp1*P1 + kd1*P2),   u2(0) = -(kp2*P1 + kd2*P2)
function test_control_initial(testCase)
    fprintf('\n--- test_control_initial ---\n');

    kp1 = 3;  kd1 = 5;
    kp2 = 10; kd2 = 10;

    p.dt    = 0.001;
    p.t_end = 0.1;
    r = run_inexact_predictor(p);

    P1 = r.P_hist(1,1);
    P2 = r.P_hist(1,2);

    u1_expected = -(kp1*P1 + kd1*P2);
    u2_expected = -(kp2*P1 + kd2*P2);

    u1_actual = r.u1_hist(1);
    u2_actual = r.u2_hist(1);

    fprintf('  u1: expected = %.10f, actual = %.10f\n', u1_expected, u1_actual);
    fprintf('  u2: expected = %.10f, actual = %.10f\n', u2_expected, u2_actual);

    err_u1 = abs(u1_actual - u1_expected);
    err_u2 = abs(u2_actual - u2_expected);

    verifyLessThan(testCase, err_u1, 1e-12, ...
        'u1(0) should match -(kp1*P1 + kd1*P2) exactly.');
    verifyLessThan(testCase, err_u2, 1e-12, ...
        'u2(0) should match -(kp2*P1 + kd2*P2) exactly.');
end

%% -----------------------------------------------------------------------
%  test_dt_refinement
%  -----------------------------------------------------------------------
%  The Euler scheme is first-order.  We check that halving dt roughly
%  halves the error relative to a fine reference solution.
function test_dt_refinement(testCase)
    fprintf('\n--- test_dt_refinement ---\n');

    t_end = 5.0;

    % Reference solution with very fine dt
    p_ref.dt    = 0.00025;
    p_ref.t_end = t_end;
    r_ref = run_inexact_predictor(p_ref);

    dt_list = [0.004, 0.002, 0.001, 0.0005];
    err = zeros(size(dt_list));

    for i = 1:length(dt_list)
        p_i.dt    = dt_list(i);
        p_i.t_end = t_end;
        r_i = run_inexact_predictor(p_i);

        % Compute max error at common time points
        % Sample at the coarsest grid of (reference, current)
        Nt_i = length(r_i.t);
        ratio = round(dt_list(i) / p_ref.dt);
        max_err = 0;
        for k = 1:Nt_i
            idx_ref = (k-1)*ratio + 1;
            if idx_ref > length(r_ref.t), break; end
            e = norm(r_i.x_hist(k,:) - r_ref.x_hist(idx_ref,:));
            if e > max_err, max_err = e; end
        end
        err(i) = max_err;
        fprintf('  dt = %.5f  =>  max error = %.6e\n', dt_list(i), err(i));
    end

    % Estimate convergence orders from successive pairs
    orders = zeros(length(dt_list)-1, 1);
    for i = 1:length(dt_list)-1
        orders(i) = log(err(i)/err(i+1)) / log(dt_list(i)/dt_list(i+1));
        fprintf('  Order (dt %.5f -> %.5f): %.3f\n', ...
                dt_list(i), dt_list(i+1), orders(i));
    end

    avg_order = mean(orders);
    fprintf('  Average convergence order: %.3f\n', avg_order);

    verifyGreaterThan(testCase, avg_order, 0.8, ...
        'Average convergence order should be at least 0.8 (Euler is O(dt)).');
end

%% -----------------------------------------------------------------------
%  test_exponential_decay
%  -----------------------------------------------------------------------
%  After the initial transient (t > 2), the log of the state norm should
%  have a negative slope, indicating exponential-like decay.
function test_exponential_decay(testCase)
    fprintf('\n--- test_exponential_decay ---\n');

    p.dt    = 0.001;
    p.t_end = 10.0;
    r = run_inexact_predictor(p);

    % Extract portion after transient
    idx_start = find(r.t >= 2.0, 1, 'first');
    t_tail = r.t(idx_start:end);
    norms  = sqrt(r.x_hist(idx_start:end, 1).^2 + r.x_hist(idx_start:end, 2).^2);

    % Remove any exactly-zero entries to avoid log issues
    valid = norms > 0;
    t_valid = t_tail(valid);
    log_norms = log(norms(valid));

    % Linear regression: log(||x||) ~ slope * t + intercept
    n = length(t_valid);
    t_mean = mean(t_valid);
    ln_mean = mean(log_norms);
    slope = sum((t_valid - t_mean) .* (log_norms - ln_mean)) / ...
            sum((t_valid - t_mean).^2);

    fprintf('  Log-norm slope for t > 2: %.4f\n', slope);

    verifyLessThan(testCase, slope, 0, ...
        'Log(||x||) slope should be negative after transient (exponential decay).');
end
